src/wallet/walletdb.cpp

```diff
@@ -835,129 +835,59 @@ void ThreadFlushWalletDB()
-                        LogPrint("db", "Flushed %s %dms\n", strFile, GetTimeMillis() - nStart);
-                    }
-                }
-            }
+            const std::string& strFile = pwalletMain->strWalletFile;
+            if (CDB::PeriodicFlush(strFile))
+                nLastFlushed = CWalletDB::GetUpdateCounter();
         }
     }
 }
 
 //
 // Try to (very carefully!) recover wallet file if there is a problem.
 //
-bool CWalletDB::Recover(CDBEnv& dbenv, const std::string& filename, bool fOnlyKeys)
-{
-    // Recovery procedure:
-    // move wallet file to wallet.timestamp.bak
-    // Call Salvage with fAggressive=true to
-    // get as much data as possible.
-    // Rewrite salvaged data to fresh wallet file
-    // Set -rescan so any missing transactions will be
-    // found.
-    int64_t now = GetTime();
-    std::string newFilename = strprintf("wallet.%d.bak", now);
-
-    int result = dbenv.dbenv->dbrename(NULL, filename.c_str(), NULL,
-                                       newFilename.c_str(), DB_AUTO_COMMIT);
-    if (result == 0)
-        LogPrintf("Renamed %s to %s\n", filename, newFilename);
-    else
-    {
-        LogPrintf("Failed to rename %s to %s\n", filename, newFilename);
-        return false;
-    }
+bool CWalletDB::Recover(const std::string& filename, void *callbackDataIn, bool (*recoverKVcallback)(void* callbackData, CDataStream ssKey, CDataStream ssValue))
+{
+    return CDB::Recover(filename, callbackDataIn, recoverKVcallback);
+}
 
-    std::vector<CDBEnv::KeyValPair> salvagedData;
-    bool fSuccess = dbenv.Salvage(newFilename, true, salvagedData);
-    if (salvagedData.empty())
+bool CWalletDB::Recover(const std::string& filename)
+{
+    // recover without a key filter callback
+    // results in recovering all record types
+    return CWalletDB::Recover(filename, NULL, NULL);
+}
+
+bool CWalletDB::RecoverKeysOnlyFilter(void *callbackData, CDataStream ssKey, CDataStream ssValue)
+{
+    CWallet *dummyWallet = reinterpret_cast<CWallet*>(callbackData);
+    CWalletScanState dummyWss;
+    std::string strType, strErr;
+    bool fReadOK;
     {
-        LogPrintf("Salvage(aggressive) found no records in %s.\n", newFilename);
-        return false;
+        // Required in LoadKeyMetadata():
+        LOCK(dummyWallet->cs_wallet);
+        fReadOK = ReadKeyValue(dummyWallet, ssKey, ssValue,
+                               dummyWss, strType, strErr);
     }
-    LogPrintf("Salvage(aggressive) found %u records\n", salvagedData.size());
-
-    std::unique_ptr<Db> pdbCopy(new Db(dbenv.dbenv, 0));
-    int ret = pdbCopy->open(NULL,               // Txn pointer
-                            filename.c_str(),   // Filename
-                            "main",             // Logical db name
-                            DB_BTREE,           // Database type
-                            DB_CREATE,          // Flags
-                            0);
-    if (ret > 0)
+    if (!IsKeyType(strType) && strType != "hdchain")
```

Comment:
Hmm, from the original PR it looks like this was a move/refactor of code. The original code removed this line with `hdchain` and re-added it with `hdchain` again. In this PR however, you're removing `hdpubkey` and re-adding `hdchain`, so you're not just moving/refactoring anymore.

I don't know why we have this difference. It is one of the many differences we have in Dash due to the way HD support was merged in back then. Maybe @UdjinM6 knows more.

Reply:
Good catch 👍 This should say "hdpubkey" indeed. Original changes https://github.com/dashpay/dash/pull/1405/commits/300e10c8717617e6d812e32609b3582d248c1c2e

Reply:
```suggestion
    if (!IsKeyType(strType) && strType != "hdpubkey")
```